db/blip_handler.go

```diff
@@ -320,14 +320,20 @@ func (bh *blipHandler) sendChanges(sender *blip.Sender, opts *sendChangesOptions
 				for _, item := range change.Changes {
 					changeRow := bh.buildChangesRow(change, item["rev"])
 
-					if bh.purgeOnRemoval && len(change.Removed) > 0 && bh.blipContext.ActiveProtocol() == BlipCBMobileReplicationV3 {
-						userHasAccessToDoc, err := UserHasDocAccess(bh.db, change.ID, item["rev"])
-						if err != nil {
-							base.InfofCtx(bh.loggingCtx, base.KeySync, "Unable to obtain the doc: %s %s to verify user access: %v", base.UD(change.ID), item["rev"], err)
-							continue
-						}
-
-						if userHasAccessToDoc {
+					if change.allRemoved && bh.blipContext.ActiveProtocol() == BlipCBMobileReplicationV3 {
+						// If client requests revocations / removals via subChanges, continue work to possibly send
+						// removal. Otherwise skip sending removals, unless its also a deletion.
+						if opts.revocations {
+							userHasAccessToDoc, err := UserHasDocAccess(bh.db, change.ID, item["rev"])
+							if err != nil {
+								base.InfofCtx(bh.loggingCtx, base.KeySync, "Unable to obtain the doc: %s %s to verify user access: %v", base.UD(change.ID), item["rev"], err)
+								continue
+							}
+
+							if userHasAccessToDoc {
+								continue
+							}
+						} else if !change.Deleted {
```

Comment:
If we're going to send tombstone removals when opts.revocations=false, should we also be sending tombstone removals when opts.revocations==true && userHasAccessToDoc == true?  (i.e. should the !change.Deleted check be part of the condition on line 323?)

Reply:
Yeah, thats a fair point. I was assuming tombstones would lose access anyway but on second thoughts that'd be a wrong assumption.
Made that change now.